metagraph/src/annotation/taxonomy/taxonomic_db.hpp

```diff
@@ -0,0 +1,162 @@
+#ifndef __TAXONOMIC_DB_HPP__
+#define __TAXONOMIC_DB_HPP__
+
+#include <tsl/hopscotch_set.h>
+#include <tsl/hopscotch_map.h>
+
+#include "annotation/representation/base/annotation.hpp"
+#include "graph/annotated_dbg.hpp"
+
+
+namespace mtg {
+namespace annot {
+
+/**
+ * TaxonomyDB constructs a taxonomic map (kmer to taxid) in a similar way
+ * to how Kraken2 works. The file exported by this object will be further
+ * used for taxonomic sequence classifier in './metagraph tax_class'.
+ */
+class TaxonomyDB {
+  public:
+    typedef std::string AccessionVersion;
+    typedef std::uint64_t TaxId;
+    typedef std::uint64_t NormalizedTaxId;
+    typedef std::vector<std::vector<NormalizedTaxId>> ChildrenList;
+    typedef annot::MultiLabelEncoded<AccessionVersion> Annotator;
+    using KmerId = Annotator::Index;
+
+  private:
+    /**
+     * node_depth returns the depth for each node.
+     * The root is the unique node with maximal depth and all the leaves have depth 1.
+     */
+    std::vector<uint64_t> node_depth;
+
+    /**
+     * rmq_data[0] contains the taxonomic tree linearization
+     *          (e.g. for root 1 and edges={1-2; 1-3}, the linearization is "1 2 1 3 1")
+     * rmq_data[l][x] returns the node with the maximal depth among positions [x, x+2^l-1]
+     * in the linearization.
+     *          (e.g. rmq_data[3][6] return the node with max depth in [6, 13])
+     */
+    std::vector<std::vector<NormalizedTaxId>> rmq_data;
+
+    /**
+     * node_to_linearization_idx[node] returns the index of the first occurrence of node
+     * in the linearization.
+     */
```

Comment:
Move private to the end. Public methods go first.

Reply:
cool, now it is in this order:

- public fields
- public methods
- private fields
- private methods